### PR TITLE
LoopIdiomRecognize: detect and convert powi idiom

### PR DESCRIPTION
**Pull request by artagnon**
The following code, when compiled under -ffast-math, produces bad codegen due to LoopVectorize:

```c
  float powi(float base, int exp) {
    float result = 1.0;
      for (int i = 0; i < exp; ++i)
        result *= base;
    return result;
  }
```

It can easily be replaced with the llvm.powi intrinsic, when the exponent is a C int type. This is the job of LoopIdiomRecognize, and has been marked as a TODO item for years. Fulfill this wish, and replace computations of this form with the llvm.powi intrinsic.

-- 8< --
Based on #72648. Please review only the second patch.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp [hunk 1]
```diff
@@ -925,6 +930,111 @@ bool LoopIdiomRecognize::processLoopMemSet(MemSetInst *MSI,
...
+        LLVM_DEBUG(dbgs() << DEBUG_TYPE " Trip count not safe to expand\n");
+        return false;
+      }
+      Value *Exp =
+          Expander.expandCodeFor(TripCount, ExpTy, PH->getTerminator());
+
+      // Insert the powi intrinsic, and replace its uses outside the block.
+      const DebugLoc &Loc = FMul->getDebugLoc();
+      Value *Base = isa<PHINode>(FMul->getOperand(1)) ? FMul->getOperand(2)
+                                                      : FMul->getOperand(1);
```
Comment (joe-img) [inline llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:1026]: Operands are off by one here - getOperand(2) will assert.

**Review by joe-img — COMMENTED** [review]
I don't know if this is out of the scope of this change, but all the loops in the unit tests get reduced to a select and the icmp 0. Would it better off eliminating the check all together? Seeing as we're calling the intrinsic and I think it can handle the 0 exponent.

File: llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp [hunk 2]
```diff
@@ -549,33 +544,43 @@ bool LoopIdiomRecognize::runOnLoopBlock(
...
-  // Look for a single store or sets of stores with a common base, which can be
-  // optimized into a memset (memset_pattern).  The latter most commonly happens
-  // with structs and handunrolled loops.
-  for (auto &SL : StoreRefsForMemset)
-    MadeChange |= processLoopStores(SL.second, BECount, ForMemset::Yes);
+  HasMemset = TLI->has(LibFunc_memset);
+  HasMemsetPattern = TLI->has(LibFunc_memset_pattern16);
+  HasMemcpy = TLI->has(LibFunc_memcpy);
+
+  if (HasMemset || HasMemsetPattern || HasMemcpy) {
```
[the remark this answers is not in the document]
Comment (artagnon) [reply llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:560]: Each function checks `HasMemset`, `HasMemsetPattern`, and `HasMemcpy`, when it needs them. This check is merely moved from the parent function to the child function.

**Comment by jcranmer-intel** [comment]
I'm skeptical of this change for quite a few reasons:

- `powi` corresponds to a GCC builtin that is only defined for C int (which is `i32` for most targets).
- `powi` isn't always lowered by targets correctly for non-`i32` parameters. Several targets fail to compile `llvm.powi.f32.i16`, for example.
- Unlike `memcpy` and `memset`, which tend to be important enough internal intrinsics that you have to handle them correctly even in freestanding modes, `powi` is... not that. Without having a way to forcibly disable the optimizer from trying to generate `powi`, I don't think this patch can even be contemplated.
- Checking for `fast` rather than appropriate individual fast-math flags is weird.
- This transform is only beneficial if the result of the multiply is *only* used outside of the loop. `replaceUsesOutsideBlock` doesn't guarantee that.
- I'm somewhat skeptical that this is going to be a useful optimization in practice. I don't see the kind of loops where this transformation would be beneficial just coming up all that frequently, even as something that might be the result of optimization.

**Comment by artagnon** [comment]
> * I'm somewhat skeptical that this is going to be a useful optimization in practice. I don't see the kind of loops where this transformation would be beneficial just coming up all that frequently, even as something that might be the result of optimization.

Addressed all other review comments. I agree that this is quite a minor optimization, and that this idiom doesn't crop up frequently; nevertheless, it is a valid optimization, and benefits some programs.

**Comment by artagnon** [comment]
Gentle ping. I think this patch would make a nice addition to LIR, even if there aren't too many idioms of this type in the wild.

**Pull request closed without merging**